Mathlib/RingTheory/FiniteType.lean

```diff
@@ -706,35 +705,114 @@ end MonoidAlgebra
 
 end MonoidAlgebra
 
+section Orzech
+
+open Submodule in
+/-- **Orzech's theorem** for finitely generated module: if `R` is a commutative ring,
+`M` and `N` are `R`-modules, `M` is finitely generated, `i : N →ₗ[R] M` is injective,
+`f : N →ₗ[R] M` is surjective, then `f` is also injective.
+
+This is a consequence of Noetherian case
+(`IsNoetherian.injective_of_surjective_of_injective`), which requires that `M` is a
+Noetherian module, but allows `R` to be non-commutative. The reduction of this result to
+Noetherian case is adapted from <https://math.stackexchange.com/a/1066110>:
+suppose `{ m_j }` is a finite set of generator of `M`, for any `n : N` one can write
+`i n = ∑ j, b_j * m_j` for `{ b_j }` in `R`.
+We can choose `{ n_j }` which are preimages of `{ m_j }` under `f`, and can choose
+`{ c_jl }` in `R` such that `i n_j = ∑ l, c_jl * m_l` for each `j`.
+Now let `A` be the subring of `R` generated by `{ b_j }` and `{ c_jl }`, then it is
+Noetherian. Let `N'` be the `A`-submodule of `N` generated by `n` and `{ n_j }`,
+`M'` be the `A`-submodule of `M` generated by `{ m_j }`,
+then it's easy to see that `i` and `f` restrict to `N' →ₗ[A] M'`,
+and the restricted version of `f` is surjective, hence by Noetherian case,
+it is also injective, in particular, if `f n = 0`, then `n = 0`.
+
+See also Orzech's original paper: *Onto endomorphisms are isomorphisms* [orzech1971]. -/
```

Comment:
This is the Orzech property! Can you add an instance?

Reply:
I have moved `commRing_orzechProperty` to this file.